devdoc/waiter_queue_requirements.md

```diff
@@ -0,0 +1,69 @@
+
+
+## Exposed API
+
+### waiter_queue_ll
+```
+#define WAITER_QUEUE_CALL_REASON_VALUES \
+    WAITER_QUEUE_CALL_REASON_POPPED, \
+    WAITER_QUEUE_CALL_REASON_ABANDONED
+
+MU_DEFINE_ENUM(WAITER_QUEUE_CALL_REASON, WAITER_QUEUE_CALL_REASON_VALUES);
+
+typedef struct WAITER_QUEUE_LL_TAG* WAITER_QUEUE_LL_HANDLE;
+typedef bool(*POP_CALLBACK)(void* context, void* data, bool* continue_processing, WAITER_QUEUE_CALL_REASON reason);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    MOCKABLE_FUNCTION(, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll_create);
+    MOCKABLE_FUNCTION(, void, waiter_queue_ll_destroy, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll);
+
+    MOCKABLE_FUNCTION(, int, waiter_queue_ll_push, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll, POP_CALLBACK, pop_callback, void*, pop_callback_context);
+    MOCKABLE_FUNCTION(, int, waiter_queue_ll_pop, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll, void*, data);
+    MOCKABLE_FUNCTION(, void, waiter_queue_ll_abandon, WAITER_QUEUE_LL_HANDLE, waiter_queue_ll);
+
+#ifdef __cplusplus
+}
+#endif
+```
+
+### waiter_queue_hl
+```
+typedef struct WAITER_QUEUE_HL_TAG* WAITER_QUEUE_HL_HANDLE;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    MOCKABLE_FUNCTION(, WAITER_QUEUE_HL_HANDLE, waiter_queue_hl_create);
+    MOCKABLE_FUNCTION(, void, waiter_queue_hl_destroy, WAITER_QUEUE_HL_HANDLE, waiter_queue_hl);
+
+    MOCKABLE_FUNCTION(, int, waiter_queue_hl_push, WAITER_QUEUE_HL_HANDLE, waiter_queue_hl, POP_CALLBACK, pop_callback, void*, pop_callback_context);
+    MOCKABLE_FUNCTION(, int, waiter_queue_hl_pop, WAITER_QUEUE_HL_HANDLE, waiter_queue_hl, void*, data);
+
+#ifdef __cplusplus
+}
+#endif
+```
+
+Consumers of data call `waiter_queue_hl_push` with a callback function and context to park a call. The call is parked in a queue and will be executed when the producers indicates that data is available by means of calling `waiter_queue_hl_pop`.
+
+Producers of data call `waiter_queue_hl_pop` with the data they would like to provide to the parked calls. If any calls had been parked by consumers by means of calling `waiter_queue_hl_push`, the parked calls are executed with the given data and `WAITER_QUEUE_CALL_REASON` as `WAITER_QUEUE_CALL_REASON_POPPED`.
+
+Calls are executed in the FIFO order in which they were parked. If the `POP_CALLBACK` for a parked call returns `true`, that call is removed from the queue. Otherwise, the parked call stays in the queue and may be executed again when more data becomes available.
+
+A single call to `waiter_queue_hl_pop` may result in executing multiple parked calls. `waiter_queue` executes each `POP_CALLBACK` in FIFO order until a `POP_CALLBACK` sets the `continue_processing` out argument to `false`.
+
+If there are any parked calls when `waiter_queue_hl_destroy` is called, their `POP_CALLBACKS` will be executed with `NULL` as data and `WAITER_QUEUE_CALL_REASON_ABONDONED` as reason.
```

Comment:
>reason [](http://example.com/codeflow?start=175&length=6)

I have read the whole md


General observations: 
1)
Traditionally data is "tranferred to", "given to", "made available to", "added", "inserted", "submitted", "posted" with `PUSH`. I think the APIs here are all reversed, because "consumers of data call push". Frankly, what they do is "pop".

Also, if these terms are confusing, it is perfectly fine to name them "produce" and "consume".

2) The term "call" is SO VAGUE here that it is better to be avoided altogether. "call" implies some imperative action - as in "do something" by means of calling a function. I don't see any function pointers or anything else even remotely similar to "call". 

Even parameter in line 49 is called "data". Or in line 19.

So at this moment I understand that there's "data". 

3) "parked" - poetically speaking - there's nothing "parked" here. To me the term "park" sounds like "interrupted" and it would necessitate a "resume", "restart" somehow.

The document doesn't explain what a "call" is conceptually or what the "parked" variety of the "call" is.

Reply:
1)Renamed the APIs: push->add_waiter, pop->unblock_waiters

The thing being queued is parked call, not the data.

2) Line 19 contains the type definition for `POP_CALLBACK`. These callbacks are provided by the consumer to be called when signaled. I've also added a definition for parked calls in the next iteration:

A "parked call" consists of a callback function and a context. The consumer provides the callback function and context when it parks a call. The consumer expects the parked call's callback function to be executed with the provided context when the producer signals.

3) Given the above definition of "parked calls", does that make things clearer? The consumer is interrupted because the data it requires to proceed is not available yet. So it "parks" execution by providing a callback function that is to be executed when signaled (told to resume) by the producer.